tests/test_flow/test_cache_api.py

```diff
@@ -77,24 +77,25 @@ def _validate_entry(self, entry):
         # instead.)
         # TODO Hmm, is the above true? On closer inspection, it looks like artifact URLs
         # are derelativized right away when we load the metadata YAML.
-        metadata_str = read_bytes_from_url(entry.metadata_url).decode("utf-8")
+        metadata_str = read_bytes_from_url(entry.metadata_url, self.gcs_fs).decode(
+            "utf-8"
+        )
         assert entry.entity in metadata_str
 
 
-def read_bytes_from_url(url):
+def read_bytes_from_url(url, gcs_fs):
```

Comment:
Oof, it's unfortunate that it's necessary to pass in `gcs_fs` here. Long-term, it seems like it would be nice if we could pass around a single object that knows how to access files for each protocol.

@namanjain Do you think it could make sense to have some sort of `MultiplexFileSystem` that wraps the others and dispatches each method to the right implementation based on the protocol? I realize that `fsspec` already has its own registry, but (a) it's global and (b) even if you could pass it around, you still have to do the work of extracting the protocol and looking up the filesystem before you can do your actual operation.

E.g., instead of this:

    protocol = protocol_from_url(url)
    fs = fsspec.filesystem(protocol)
    return fs.cat_file(url)

we could just do this:

    return mux_fs.cat_file(url)

Reply:
I think that makes sense to use this mux fs in test code. But we should be careful about what we store in that mux fs since we will probably send it across to subprocesses for parallel processing.

Reply:
I agree, we'd want the mux FS to be easily transmittable to other processes. But is that an argument against using it for non-test code? It seems like the alternative would be to use two separate fsspec filesystems -- local and GCS -- and both of those would need to be transmitted as well, so grouping them together doesn't seem like an added cost. I had imagined that we could use this mux FS as a single interface for accessing the disk and GCS.

Reply:
Yeah, it doesn't seem too bad for now. I was a bit worried that future fs we use might not be serializable but I guess we can have custom pickle code in mux fs for non-pickleable fs.

Reply:
It sounds like fsspec has [serialisability](https://filesystem-spec.readthedocs.io/en/latest/features.html#serialisability) as an explicit goal, so hopefully this won't be a problem with any popular filesystems.

Reply:
TIL and that's awesome!